test/services/test_gandi_service.py

```diff
@@ -13,20 +13,59 @@ def test_valid_certificate_list_returns_expected_domain(self):
         response = GandiService(test_config, 'api_key', 'base_url', 'url_ext') \
             .get_expired_certificates_from_valid_certificate_list(
             TestData.generate_filtered_certificate_list_with_expiry_date(30),
-            TestData.generate_valid_email_list())
-        self.assertIn(TestData.test_domain_name, response)
+            TestData.generate_single_email_list())
+        self.assertIn(TestData.test_domain_name_root, response)
 
 
 @freeze_time("2023-01-01")
-class TestValidCertificateRetrieval(unittest.TestCase):
+class TestCertificateRetrievalValidity(unittest.TestCase):
     def test_only_valid_certificates_are_returned_from_list(self):
         response = GandiService(test_config, 'api_key', 'base_url', 'url_ext') \
             .get_certificates_in_valid_state(
-            TestData.generate_gandi_certificate_state('valid'),
-            TestData.generate_valid_email_list()
+            TestData.generate_single_gandi_certificate_state('valid'),
+            TestData.generate_single_email_list()
         )
 
-        self.assertIn(TestData.test_domain_name, response)
+        self.assertIn(TestData.test_domain_name_root, response)
+
+    def test_invalid_certificate_is_not_returned(self):
+        response = GandiService(test_config, 'api_key', 'base_url', 'url_ext') \
+            .get_certificates_in_valid_state(
+            TestData.generate_single_gandi_certificate_state('pending'),
+            TestData.generate_single_email_list()
+        )
+
+        self.assertNotIn(TestData.test_domain_name_root, response)
+
+    def test_multiple_valid_certificates_returned(self):
+        test_case_count = 3
+        response = GandiService(test_config, 'api_key', 'base_url', 'url_ext') \
+            .get_certificates_in_valid_state(
+            TestData.generate_multiple_gandi_certificate_states(
+                'valid', test_case_count),
+            TestData.generate_multiple_email_list(test_case_count)
+        )
+        for i in range(test_case_count):
```

Comment:
Best avoiding using logic during the tests such as `for` loops. The reason is that the tests are there to test the logic, so if there's logic in the tests, then who's testing the tests! 🧪 

I believe the `assertIn()` should cover the functionality you need here, though - i.e. something like 
```python
self.assertIn([TestData.test_domain_name_root + "0", TestData.test_domain_name_root + "1", TestData.test_domain_name_root + "2"], response)
```

or if that doesn't work

```python
self.assertIn(TestData.test_domain_name_root + "0", response)
self.assertIn(TestData.test_domain_name_root + "1", response)
self.assertIn(TestData.test_domain_name_root + "2", response)
```

Doesn't look as clean, but a nicer standard to reduce the complexity when debugging tests 😄

Reply:
Cool beans, I'll move everything out to separate asserts 🙌